tfx/experimental/pipeline_testing/pipeline_recorder_utils.py

```diff
@@ -0,0 +1,157 @@
+# Lint as: python3
+# Copyright 2020 Google LLC. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Recording pipeline from MLMD metadata."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+from collections import defaultdict
+import os
+
+from absl import logging
+from ml_metadata.proto import metadata_store_pb2
+import tensorflow as tf
+from typing import Iterable, List, Mapping, Optional, Text, Tuple
+
+from tfx.orchestration import metadata
+from tfx.utils import io_utils
+
+def _get_paths(metadata_connection: metadata.Metadata,
+               execution_ids: List[int],
+               output_dir: Text) -> Iterable[Tuple[Text, Text]]:
+  """Returns an iterable of tuple containing source artifact uris and
+  destination uris, which are located in the output_dir. The source
+  artifact uris are retrieved using execution ids.
+
+  Args:
+    metadata_connection: Instance of metadata.Metadata for I/O to MLMD.
+    execution_ids: List of execution ids of a pipeline run.
+    output_dir: Directory path where the pipeline outputs should be recorded.
+
+  Returns:
```

Comment:
Should be 'Yields' for generators.

https://google.github.io/styleguide/pyguide.html#doc-function-returns